### ISSUE TITLE
AST error reported

### ISSUE THREAD

**Issue by recursion-ninja** [tweag/ormolu#874]
**Describe the bug**
Abstract syntax tree of input and formatted code differ (reportedly).

**To Reproduce**
```
$ curl -sSL \
https://raw.githubusercontent.com/recursion-ninja/bv-little/bdc7c746faa663a6f15fbd462f3f915f7f5ef7e3/src/core/Data/BitVector/LittleEndian/Internal.hs \
| ormolu
```
Output:
```
AST of input and AST of formatted code differ.
Please, consider reporting the bug.
```

**Expected behavior**
Formats the file, printing the formatted output to STDOUT rather that encounter and then report an internal error

**Environment**
 - OS name + version: 
 ```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.4 LTS
Release:	20.04
Codename:	focal
 ```
 - Version of the code:
```
$ ormolu --version
ormolu 0.4.0.0 UNKNOWN UNKNOWN
using ghc-lib-parser 9.2.2.20220307
```

**Comment by amesgen** [comment]
Thanks for the report! I can't reproduce this, neither with Ormolu 0.4.0.0 nor with [Ormolu Live](https://ormolu-live.tweag.io/) (which tracks the latest commit to `master`), formatting your example file works fine, also with idempotency checking enabled.

How did you produce your Ormolu binary? Can you also try with the [prebuilt ones](https://github.com/tweag/ormolu/releases/tag/0.4.0.0) (it doesn't seem like you are using them, as your version output is missing the Git information which is present in the prebuilt ones)?

**Comment by recursion-ninja** [comment]
@amesgen I accidentally used a reference to the wrong branch when specifying the triggering file. I have edited the original issue description with the correct file reference.

**Comment by amesgen** [comment]
Thanks, can now reproduce on 0.4.0.0. Minimal example:

```haskell
data A = A
    { a :: A
      -- ^ foo
    }
    -- ^ bar
    -- ^ baz
```
which was formatted (with `--unsafe`) to
```haskell
data A = A
  { -- | foo
    a :: A
  }

-- bar
-- ^ baz
```

 It works on master though, and indeed, this has been fixed as a side effect of #837 (which introduced a more principled approach to format invalid Haddock comments). Concretely, this is now formatted to
```haskell
data A = A
  { -- | foo
    a :: A
  }

-- \^ bar
-- \^ baz
```

So formatting your file will work with the next Ormolu release.

**Comment by amesgen** [comment]
In your file, this manifests e.g. in the `@since` comments here:

```haskell
-- |
-- A little-endian bit vector of non-negative dimension.
data BitVector 
    = BV
    { dim :: {-# UNPACK #-} !Word
      -- ^ The /dimension/ of a bit vector.
    , nat :: !Natural
      -- ^ The value of a bit vector, as a natural number.
    }
    -- ^ @since 0.1.0
    -- ^ @since 0.1.0
    -- ^ @since 0.1.0
    deriving stock (Data, Generic)
```
